// @cloudbase/framework-core/lib/plugin-service-api/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const toolbox_1 = require("@cloudbase/toolbox");
const link_1 = require("../utils/link");
const generator_1 = require("../generator");
const api_1 = require("../api");
const emoji_1 = require("../utils/emoji");
const spawn_1 = require("../utils/spawn");
/**
 * 插件服务注入为插件提供的 API
 */
class PluginServiceApi {
    constructor(pluginManager) {
        this.pluginManager = pluginManager;
    }
    /**
     * 获取 manager 实例
     */
    get cloudbaseManager() {
        return this.pluginManager.context.cloudbaseManager;
    }
    /**
     * 获取项目根路径
     */
    get projectPath() {
        return this.pluginManager.context.projectPath;
    }
    /**
     * 项目环境id
     */
    get envId() {
        return this.pluginManager.context.envId;
    }
    /**
     * 项目地域信息
     */
    get region() {
        return this.pluginManager.context.cloudbaseConfig.region;
    }
    /**
     * 是否产生新版本
     */
    get bumpVersion() {
        return this.pluginManager.context.bumpVersion;
    }
    /**
     * 新版本的备注信息
     */
    get versionRemark() {
        return this.pluginManager.context.versionRemark;
    }
    /**
     * 获取当前账号 API 密钥信息
     */
    get getCredential() {
        return toolbox_1.checkAndGetCredential;
    }
    /**
     * secretId
     * @deprecated
     */
    get secretId() {
        return this.pluginManager.context.cloudbaseConfig.secretId || '';
    }
    /**
     * secretKey
     * @deprecated
     */
    get secretKey() {
        return this.pluginManager.context.cloudbaseConfig.secretKey || '';
    }
    /**
     * token
     * @deprecated
     */
    get token() {
        return this.pluginManager.context.cloudbaseConfig.token || '';
    }
    /**
     * 获取 logger 实例
     */
    get logger() {
        return this.pluginManager.context.logger;
    }
    /**
     * 获取资源操作 API
     */
    get resourceProviders() {
        return this.pluginManager.context.resourceProviders;
    }
    /**
     * 获取云开发 CLI 配置文件
     */
    get projectConfig() {
        return this.pluginManager.context.projectConfig;
    }
    /**
     * 获取generator实例
     */
    get generator() {
        return new generator_1.Generator();
    }
    /**
     * 云API
     */
    get cloudApi() {
        return api_1.CloudApi;
    }
    /**
     * 生成命令行链接
     */
    get genClickableLink() {
        return link_1.genClickableLink;
    }
    /**
     * 安全地使用emoji
     */
    get emoji() {
        return emoji_1.emoji;
    }
    /**
     * 获取 cloudbase 配置
     */
    get appConfig() {
        return this.pluginManager.context.appConfig;
    }
    /**
     * 获取 samManager API
     */
    get samManager() {
        return this.pluginManager.context.samManager;
    }
    /**
     * 获取构建的CIId
     */
    get ciId() {
        return this.pluginManager.context.ciId;
    }
    /**
     * 调用子进程
     */
    get spawnPromise() {
        return spawn_1.spawnPromise;
    }
}
exports.default = PluginServiceApi;
