// @cloudbase/framework-core/lib/context/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const manager_node_1 = __importDefault(require("@cloudbase/manager-node"));
const logger_1 = __importDefault(require("../logger"));
class Context {
    constructor({ appConfig, cloudbaseConfig, projectPath, projectConfig, logLevel, resourceProviders, samManager, bumpVersion, versionRemark, ciId, }) {
        this.appConfig = appConfig;
        this.projectPath = projectPath;
        this.cloudbaseManager = new manager_node_1.default(cloudbaseConfig);
        this.cloudbaseConfig = cloudbaseConfig;
        this.envId = cloudbaseConfig.envId;
        this.logger = (0, logger_1.default)(logLevel);
        this.resourceProviders = resourceProviders;
        this.projectConfig = projectConfig;
        this.samManager = samManager;
        this.bumpVersion = bumpVersion;
        this.versionRemark = versionRemark;
        this.ciId = ciId;
    }
}
exports.default = Context;
