// @cloudbase/framework-core/lib/api/domain.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchDomains = exports.describeCloudBaseGWService = exports.describeStaticRes = void 0;
/**
 * Tencent is pleased to support the open source community by making CloudBaseFramework - 云原生一体化部署工具 available.
 *
 * Copyright (C) 2020 THL A29 Limited, a Tencent company.  All rights reserved.
 *
 * Please refer to license text included with this package for license details.
 */
const _1 = require(".");
function describeStaticRes() {
    return __awaiter(this, void 0, void 0, function* () {
        return _1.CloudApi.tcbService.request('DescribeStaticStore', {});
    });
}
exports.describeStaticRes = describeStaticRes;
function describeCloudBaseGWService() {
    return __awaiter(this, void 0, void 0, function* () {
        return _1.CloudApi.tcbUinService.request('DescribeCloudBaseGWService', {
            ServiceId: _1.CloudApi.envId,
            EnableRegion: true,
            EnableUnion: true,
        });
    });
}
exports.describeCloudBaseGWService = describeCloudBaseGWService;
function fetchDomains() {
    var _a, _b;
    return __awaiter(this, void 0, void 0, function* () {
        const gwServiceInfo = yield describeCloudBaseGWService();
        const staticResInfo = yield describeStaticRes();
        return {
            static: (_b = (_a = staticResInfo === null || staticResInfo === void 0 ? void 0 : staticResInfo.Data) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.CdnDomain,
            service: gwServiceInfo === null || gwServiceInfo === void 0 ? void 0 : gwServiceInfo.DefaultDomain,
        };
    });
}
exports.fetchDomains = fetchDomains;
