// @cloudbase/framework-plugin-function/lib/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.plugin = exports.FunctionBuilder = void 0;
/**
 * Tencent is pleased to support the open source community by making CloudBaseFramework - 云原生一体化部署工具 available.
 *
 * Copyright (C) 2020 THL A29 Limited, a Tencent company.  All rights reserved.
 *
 * Please refer to license text included with this package for license details.
 */
const path_1 = __importDefault(require("path"));
const archiver_1 = __importDefault(require("archiver"));
const fs_1 = __importDefault(require("fs"));
const framework_core_1 = require("@cloudbase/framework-core");
const toolbox_1 = require("@cloudbase/toolbox");
const lodash_merge_1 = __importDefault(require("lodash.merge"));
class FunctionPlugin extends framework_core_1.Plugin {
    constructor(name, api, inputs) {
        super(name, api, inputs);
        this.name = name;
        this.api = api;
        this.inputs = inputs;
        const config = this.api.projectConfig;
        const DEFAULT_INPUTS = {
            functionRootPath: (config === null || config === void 0 ? void 0 : config.functionRoot) || 'cloudfunctions',
            functions: config === null || config === void 0 ? void 0 : config.functions,
            servicePaths: {},
            functionDefaultConfig: config === null || config === void 0 ? void 0 : config.functionDefaultConfig,
        };
        this.resolvedInputs = resolveInputs(this.inputs, DEFAULT_INPUTS);
        let publishIncludeList;
        if (this.resolvedInputs.publishIncludeList) {
            this.api.logger.debug('publishIncludeList', this.resolvedInputs.publishIncludeList);
            publishIncludeList = this.resolvedInputs.publishIncludeList.split(',');
        }
        this.resolvedInputs.functions = this.resolvedInputs.functions
            .filter((func) => {
            if (publishIncludeList) {
                return publishIncludeList.includes(func.name);
            }
            else {
                return true;
            }
        })
            .map((func) => {
            return (0, lodash_merge_1.default)({}, {
                runtime: 'Nodejs10.15',
                installDependency: true,
                handler: 'index.main',
            }, this.resolvedInputs.functionDefaultConfig, func);
        });
        this.functions = this.resolvedInputs.functions;
        this.functionRootPath = path_1.default.isAbsolute(this.resolvedInputs.functionRootPath)
            ? this.resolvedInputs.functionRootPath
            : path_1.default.join(this.api.projectPath, this.resolvedInputs.functionRootPath);
        this.builder = new FunctionBuilder({
            projectPath: this.api.projectPath,
        });
        this.outputs = {};
    }
    /**
     * 初始化
     */
    init() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            this.api.logger.debug('FunctionPlugin: init', this.resolvedInputs);
            if (!((_a = this.functions) === null || _a === void 0 ? void 0 : _a.length)) {
                throw new Error('云函数插件配置有误，函数列表为空');
            }
        });
    }
    compile() {
        return __awaiter(this, void 0, void 0, function* () {
            this.api.logger.debug('FunctionPlugin: compile', this.resolvedInputs);
            const builderOptions = this.functions.map((func) => {
                var _a, _b;
                let fileName = func.name;
                let localFunctionPath;
                let isNeedZip = true;
                if ((_a = func.runtime) === null || _a === void 0 ? void 0 : _a.includes('Java')) {
                    fileName = func.name + '.jar';
                    isNeedZip = false;
                }
                if (func.functionDistPath) {
                    localFunctionPath = path_1.default.join(this.functionRootPath, func.functionDistPath);
                }
                else {
                    localFunctionPath = path_1.default.join(this.functionRootPath, fileName);
                }
                if (((_b = func.runtime) === null || _b === void 0 ? void 0 : _b.includes('Node')) && func.installDependency) {
                    const packageJSONExists = fs_1.default.existsSync(path_1.default.join(localFunctionPath, 'package.json'));
                    if (!packageJSONExists) {
                        this.api.logger.warn(`函数 ${func.name} 设置了云端安装依赖，但函数代码根目录下未提供 package.json`);
                        func.installDependency = false;
                    }
                }
                const zipName = `${func.name + Date.now()}.zip`;
                return {
                    name: func.name,
                    localPath: localFunctionPath,
                    zipFileName: zipName,
                    isNeedZip,
                    ignore: func.installDependency
                        ? ['node_modules/**/*', 'node_modules', '.git/**', ...(func.ignore || [])]
                        : [...(func.ignore || [])],
                };
            });
            const buildResult = yield this.builder.build(builderOptions);
            const codeUris = yield this.api.samManager.uploadFile(buildResult.functions.map((func) => {
                return {
                    fileType: 'FUNCTION',
                    fileName: `${func.name}.zip`,
                    filePath: func.source,
                };
            }));
            buildResult.functions.forEach((func, index) => {
                this.outputs[func.name] = codeUris[index];
            });
            let a = {};
            return {
                EntryPoint: Object.values(this.resolvedInputs.servicePaths).map((servicePath) => {
                    return {
                        Label: '服务地址',
                        EntryType: 'HttpService',
                        HttpEntryPath: servicePath,
                    };
                }),
                Resources: this.functions.reduce((resources, func) => {
                    resources[this.toConstantCase(func.name)] = this.functionConfigToSAM(func);
                    return resources;
                }, a),
            };
        });
    }
    /**
     * 执行本地命令
     */
    run() {
        return __awaiter(this, void 0, void 0, function* () { });
    }
    /**
     * 删除资源
     */
    remove() {
        return __awaiter(this, void 0, void 0, function* () { });
    }
    /**
     * 生成代码
     */
    genCode() {
        return __awaiter(this, void 0, void 0, function* () { });
    }
    /**
     * 构建
     */
    build() {
        return __awaiter(this, void 0, void 0, function* () {
            this.api.logger.debug('FunctionPlugin: build', this.resolvedInputs);
        });
    }
    /**
     * 部署
     */
    deploy() {
        return __awaiter(this, void 0, void 0, function* () {
            this.api.logger.debug('FunctionPlugin: deploy', this.resolvedInputs, this.buildOutput);
            // 批量部署云函数
            yield Promise.all(this.functions.map((func) => __awaiter(this, void 0, void 0, function* () {
                this.api.logger.info(`${this.api.emoji('🚀')} [${func.name}] 云函数部署成功`);
            })));
            this.api.logger.info(`${this.api.emoji('🚀')} 云函数部署成功`);
        });
    }
    functionConfigToSAM(functionConfig) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        const networkConfig = this.api.appConfig.network;
        return Object.assign({
            Type: 'CloudBase::Function',
            Properties: Object.assign({
                Handler: functionConfig.handler || 'index.main',
                Description: '无服务器执行环境，帮助您在无需购买和管理服务器的情况下运行代码',
                Runtime: functionConfig.runtime,
                FunctionName: functionConfig.name,
                MemorySize: functionConfig.memorySize || functionConfig.memory || 128,
                Timeout: functionConfig.timeout || 5,
                Environment: {
                    Variables: normalizeEnvironments(functionConfig.envVariables),
                },
                VpcConfig: {
                    VpcId: ((_a = functionConfig.vpc) === null || _a === void 0 ? void 0 : _a.vpcId) ||
                        ((networkConfig === null || networkConfig === void 0 ? void 0 : networkConfig.uniqVpcId)
                            ? '${Outputs.Network.Properties.InstanceId}'
                            : undefined),
                    SubnetId: (_b = functionConfig.vpc) === null || _b === void 0 ? void 0 : _b.subnetId,
                },
                HttpPath: this.resolvedInputs.servicePaths[functionConfig.name],
                InstallDependency: ((_c = functionConfig.runtime) === null || _c === void 0 ? void 0 : _c.includes('Node')) &&
                    'installDependency' in functionConfig
                    ? functionConfig.installDependency
                    : false,
                CodeUri: (_d = this.outputs[functionConfig.name]) === null || _d === void 0 ? void 0 : _d.codeUri,
                CodeSecret: !!functionConfig.codeSecret,
                Role: 'TCB_QcsRole',
                PublicNetStatus: functionConfig.publicNet,
                EipStatus: functionConfig.eip,
            }, ((_e = this.resolvedInputs.serviceConfig) === null || _e === void 0 ? void 0 : _e[functionConfig.name]) && {
                HttpPath: (_f = this.resolvedInputs.serviceConfig) === null || _f === void 0 ? void 0 : _f[functionConfig.name].httpPath,
                HttpPathEnableAuth: ((_g = this.resolvedInputs.serviceConfig) === null || _g === void 0 ? void 0 : _g[functionConfig.name].httpPathEnableAuth) || false
            }, (this.api.bumpVersion || functionConfig.bumpVersion) && {
                NewVersion: true,
            }, this.api.versionRemark && {
                VersionRemark: this.api.versionRemark,
            }, functionConfig.aclRule && {
                AclTag: 'CUSTOM',
                AclRule: this.genAclRule(functionConfig),
            }, ((_h = functionConfig.triggers) === null || _h === void 0 ? void 0 : _h.length) && {
                Events: functionConfig.triggers.reduce((prev, cur) => {
                    prev[cur.name] = {
                        Type: 'Timer',
                        Properties: {
                            CronExpression: cur.config,
                            Message: cur.name,
                            Enable: true,
                        },
                    };
                    return prev;
                }, {}),
            }),
        });
    }
    toConstantCase(name) {
        let result = '';
        let lastIsDivide = true;
        for (let letter of name) {
            if (letter === '-' || letter === '_') {
                lastIsDivide = true;
            }
            else if (lastIsDivide) {
                result += letter.toUpperCase();
                lastIsDivide = false;
            }
            else {
                result += letter.toLowerCase();
                lastIsDivide = false;
            }
        }
        return result;
    }
    genAclRule(functionConfig) {
        const aclRule = {};
        aclRule[functionConfig.name] = functionConfig.aclRule;
        return JSON.stringify(aclRule);
    }
}
function resolveInputs(inputs, defaultInputs) {
    return Object.assign({}, defaultInputs, inputs);
}
class FunctionBuilder extends framework_core_1.Builder {
    constructor(options) {
        super(Object.assign({ type: 'function' }, options));
    }
    build(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const functions = yield Promise.all(options.map((option) => __awaiter(this, void 0, void 0, function* () {
                let source;
                if (!fs_1.default.existsSync(this.distDir)) {
                    (0, toolbox_1.mkdirSync)(this.distDir);
                }
                if (!fs_1.default.existsSync(option.localPath)) {
                    throw new Error(`函数目录或者文件 ${path_1.default.basename(option.localPath)} 不存在`);
                }
                const fileStats = fs_1.default.statSync(option.localPath);
                if (!option.isNeedZip) {
                    source = option.localPath;
                }
                else {
                    const localZipPath = path_1.default.join(this.distDir, option.zipFileName);
                    source = localZipPath;
                    if (fileStats.isFile()) {
                        this.logger.debug('option.localPath', option.localPath, localZipPath);
                        yield this.zipFile(option.localPath, localZipPath);
                    }
                    else if (fileStats.isDirectory()) {
                        this.logger.debug('option.localPath', option.localPath, localZipPath);
                        yield this.zipDir(option.localPath, localZipPath, option.ignore);
                    }
                }
                return {
                    name: option.name,
                    options: {},
                    source,
                    entry: option.zipFileName,
                };
            })));
            return {
                functions,
            };
        });
    }
    zipFile(src, dest) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                // create a file to stream archive data to.
                var output = fs_1.default.createWriteStream(dest);
                var archive = (0, archiver_1.default)('zip', {
                    zlib: { level: 9 }, // Sets the compression level.
                });
                output.on('close', () => {
                    resolve(void 0);
                });
                archive.on('error', reject);
                archive.file(src, {
                    name: path_1.default.basename(src),
                });
                archive.pipe(output);
                archive.finalize();
            });
        });
    }
    zipDir(src, dest, ignore) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, reject) => {
                // create a file to stream archive data to.
                var output = fs_1.default.createWriteStream(dest);
                var archive = (0, archiver_1.default)('zip', {
                    zlib: { level: 9 }, // Sets the compression level.
                });
                output.on('close', () => {
                    resolve(void 0);
                });
                archive.on('error', reject);
                archive.glob('**/*', {
                    cwd: src,
                    ignore: ignore || [],
                    dot: true,
                }, {});
                archive.pipe(output);
                archive.finalize();
            });
        });
    }
}
exports.FunctionBuilder = FunctionBuilder;
function normalizeEnvironments(envVariables = {}) {
    if (!envVariables)
        return {};
    let result = {};
    for (let i in envVariables) {
        let value = envVariables[i];
        if (value) {
            result[i] = String(value);
        }
    }
    return result;
}
exports.plugin = FunctionPlugin;
