// @cloudbase/framework-core/lib/sam/api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SamApi = void 0;
/**
 * Tencent is pleased to support the open source community by making CloudBaseFramework - 云原生一体化部署工具 available.
 *
 * Copyright (C) 2020 THL A29 Limited, a Tencent company.  All rights reserved.
 *
 * Please refer to license text included with this package for license details.
 */
const api_1 = require("../api");
class SamApi {
    constructor() { }
    /**
     *
     * 创建私有版本并部署
     * @param template
     * 返回
     * ExtensionId	String	扩展ID
     * RequestId	String	唯一请求 ID，每次请求都会返回。定位问题时需要提供该次请求的 RequestId。
     */
    createAndInstall(template) {
        return api_1.CloudApi.tcbService.request('CreatePrivateExtensionAndInstall', {
            Template: template,
        });
    }
    /**
     * 查询扩展任务的状态
     */
    fetchExtensionTaskStatus(ids) {
        return api_1.CloudApi.tcbService.request('DescribeExtensionTaskStatus', {
            ExtensionIds: ids,
        });
    }
    /**
     * 上报部署状态
     */
    reportCloudBaseCIResultCallback(ciId, traceId, extensionId) {
        return api_1.CloudApi.tcbService.request('CloudBaseCIResultCallback', {
            CIID: ciId,
            TraceId: traceId,
            ExtensionID: extensionId,
        });
    }
    /**
     * 描述扩展上传文件信息
     */
    describeExtensionUploadInfo(files) {
        return api_1.CloudApi.tcbUinService.request('DescribeExtensionUploadInfo', {
            ExtensionFiles: files,
        });
    }
}
exports.SamApi = SamApi;
