// @cloudbase/toolbox/lib/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudBaseError = void 0;
class CloudBaseError extends Error {
    constructor(message, options = {}) {
        super();
        this.name = 'CloudBaseError';
        const { code = '', action = '', original = null, requestId = '', source } = options;
        this.options = options;
        this.message = this.getFormatMessage(message);
        this.original = original;
        this.code = code;
        this.requestId = requestId;
        this.action = action;
        this.source = source;
    }
    getFormatMessage(message) {
        const { code = '', action = '', original = null, requestId = '' } = this.options;
        const msg = [];
        if (action) {
            msg.push(`[${action}]`);
        }
        if (requestId) {
            msg.push(`RequestId: ${requestId}`);
        }
        if (code) {
            msg.push(`Error Code: ${code}`);
        }
        if (message) {
            msg.push(message);
        }
        return msg.join('，');
    }
}
exports.CloudBaseError = CloudBaseError;
